staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/established_controller.go

```diff
@@ -0,0 +1,142 @@
+type EstablishingController struct {
+	client *internalclientset.Clientset
+	queue  workqueue.RateLimitingInterface
+}
+
+// NewEstablishingController creates new EstablishingController.
+func NewEstablishingController(client *internalclientset.Clientset) *EstablishingController {
+	return &EstablishingController{
+		client: client,
+		queue:  workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "crdEstablishing"),
+	}
+}
+
+func (ec *EstablishingController) Run(stopCh <-chan struct{}) {
+	defer utilruntime.HandleCrash()
+	defer ec.queue.ShutDown()
+
+	glog.Infof("Starting EstablishingController")
+	defer glog.Infof("Shutting down EstablishingController")
+
+	// only start one worker thread since its a slow moving API
+	go wait.Until(ec.runWorker, time.Second, stopCh)
+
+	<-stopCh
+}
+
+func (ec *EstablishingController) runWorker() {
+	for ec.processNextWorkItem() {
+	}
+}
+
+// processNextWorkItem deals with one key off the queue.
+// It returns false when it's time to quit.
+func (ec *EstablishingController) processNextWorkItem() bool {
+	key, quit := ec.queue.Get()
+	if quit {
+		return false
+	}
+	defer ec.queue.Done(key)
+
+	err := ec.establishCRD(key.(string))
+	if err == nil {
+		return true
+	}
+
+	utilruntime.HandleError(fmt.Errorf("%v failed with: %v", key, err))
+	ec.queue.Add(key)
+
+	return true
+}
+
+// EstablishCRD is used to turn CRDs into the Established state.
+func (ec *EstablishingController) establishCRD(key string) error {
+	// Poll because if multiple servers try to change CRD (e.g. in HA) this can fail.
+	// That failure is normal, and so, we're trying to set it to Established again.
+	crd, err := ec.client.Apiextensions().CustomResourceDefinitions().Get(key, v1.GetOptions{})
+	if err != nil {
+		return err
+	}
+
+	if apiextensions.IsCRDConditionTrue(crd, apiextensions.Installing) {
+		establishedCondition := apiextensions.CustomResourceDefinitionCondition{
+			Type:    apiextensions.Established,
+			Status:  apiextensions.ConditionTrue,
+			Reason:  "InitialNamesAccepted",
+			Message: "the initial names have been accepted",
+		}
+
+		// Set Installing to false as it's Established now, with appropriate message depending on
+		// are we Establishing CRD for the first time or we are updating it.
+		if !apiextensions.IsCRDConditionTrue(crd, apiextensions.Established) {
+			installingCondition := apiextensions.CustomResourceDefinitionCondition{
+				Type:    apiextensions.Installing,
+				Status:  apiextensions.ConditionFalse,
+				Reason:  "CRDEstablished",
+				Message: "the crd have been established",
+			}
+			apiextensions.SetCRDCondition(crd, installingCondition)
+		} else {
+			installingCondition := apiextensions.CustomResourceDefinitionCondition{
+				Type:    apiextensions.Installing,
+				Status:  apiextensions.ConditionFalse,
+				Reason:  "CRDUpdated",
+				Message: "the crd have been updated",
+			}
+			apiextensions.SetCRDCondition(crd, installingCondition)
```

Comment:
would condense those two branches even more. Just create the condition first, and override the two differing fields in the one case.